pkgs/development/python-modules/torch/default.nix

```diff
@@ -49,9 +55,7 @@ let
   inherit (cudaPackages) cudatoolkit cudaFlags cudnn nccl;
 in
 
-# assert that everything needed for cuda is present and that the correct cuda versions are used
-assert !cudaSupport || (let majorIs = lib.versions.major cudatoolkit.version;
-                        in majorIs == "9" || majorIs == "10" || majorIs == "11");
+assert cudaSupport -> (cudaPackages.cudaMajorVersion == "11");
```

Comment:
might be worth adding a comment with a link/reference explaining that pytorch is not yet compatible with cuda 12

Reply:
I didn't actually check whether they support cuda 12. The release notes do mention that "the minimum version is 11". So I'm basically just refactoring `!... || (...)` into `... -> ...`, and then I'm removing the explicitly unsupported versions.

Reply:
Upstream doesn't specify version constraints in CMakeLists.txt when looking up `CUDAToolkit`: https://github.com/pytorch/pytorch/blob/5f57b363184cbb330551cc7fdee820b7067ac6fd/cmake/public/cuda.cmake#L57
They do assert the `>= 11` upperbound: https://github.com/pytorch/pytorch/blob/5f57b363184cbb330551cc7fdee820b7067ac6fd/cmake/public/cuda.cmake#L71

I assume pytorch should work with cuda 12 now, so we can have a follow-up PR bumping the default `cudaPackages` major version. Since this would involve a lot of extra testing, I would leave the current PR as is.

Tracking in https://github.com/NixOS/nixpkgs/issues/222778

Reply:
There's a slight issue with memory allocation due to an API change, IIRC it's the last hurdle to releasing CUDA 12 wheels: https://github.com/pytorch/pytorch/issues/91122.